deepmd/utils/learning_rate.py

```diff
@@ -105,3 +105,174 @@ def start_lr(self) -> float:
+                self.cd["decay_steps"]
+                if self.cd["decay_steps"] is not None
+                else stop_step
+            )
+
+        return tf.train.cosine_decay(
+            self.start_lr_,
+            global_step,
+            self.decay_steps_,
+            self.alpha_,
+            name="cosine",
+        )
+
+    def start_lr(self) -> float:
+        """Get the start lr."""
+        return self.start_lr_
+
+    def value(self, step: int) -> float:
+        """Get the lr at a certain step."""
+        step = min(step, self.decay_steps_)
+        cosine_decay = 0.5 * (1 + np.cos(np.pi * step / self.decay_steps_))
+        decayed = (1 - self.alpha_) * cosine_decay + self.alpha_
+        decayed_learning_rate = self.start_lr_ * decayed
+        return decayed_learning_rate
+
+
+class LearningRateCosRestarts:
+    r"""The cosine decaying restart learning rate.
+
+  The function returns the cosine decayed learning rate while taking into account
+  possible warm restarts.
+  ```
+
+    Parameters
+    ----------
+    start_lr
+            Starting learning rate 
+    stop_lr
+            Minimum learning rate value as a fraction of learning_rate.
+    decay_steps
+            Number of steps to decay over.
+    """
+
+    def __init__(
+        self,
+        start_lr: float,
+        stop_lr: float = 5e-8,
+        decay_steps: int = 10000,
+    ) -> None:
+        """Constructor."""
+        self.cd = {}
+        self.cd["start_lr"] = start_lr
+        self.cd["stop_lr"] = stop_lr
+        self.cd["decay_steps"] = decay_steps
+        self.start_lr_ = self.cd["start_lr"]
+        self.alpha_ = self.cd["stop_lr"]/self.cd["start_lr"]
+
+    def build(
+        self, global_step: tf.Tensor, stop_step: Optional[int] = None
+    ) -> tf.Tensor:
+        """Build the learning rate.
+
+        Parameters
+        ----------
+        global_step
+            The tf Tensor prividing the global training step
+        stop_step
+            The stop step.
+
+        Returns
+        -------
+        learning_rate
+            The learning rate
+        """
+        if stop_step is None:
+            self.decay_steps_ = (
+                self.cd["decay_steps"] if self.cd["decay_steps"] is not None else 10000
+            )
+        else:
+            self.stop_lr_ = (
+                self.cd["stop_lr"] if self.cd["stop_lr"] is not None else 5e-8
+            )
+            self.decay_steps_ = (
+                self.cd["decay_steps"]
+                if self.cd["decay_steps"] is not None
+                else stop_step
+            )
+
+ 
+
+        return tf.train.cosine_decay_restarts(
+            learning_rate=self.start_lr_,
+            global_step=global_step,
+            first_decay_steps=self.decay_steps_,
+            alpha=self.alpha_,
+            name="cosinerestart",
+        )
+
+    def start_lr(self) -> float:
+        """Get the start lr."""
+        return self.start_lr_
+
+    def value(self, step: int) -> float:
```

Comment:
you may not need to implement the value method if you do not print the information regarding the learning rate at the beginning of the training:
https://github.com/hellozhaoming/deepmd-kit/blob/05052c195308f61b63ce2bab130ce0e8cba60604/deepmd/train/trainer.py#L566